File: frontend/src/types/query.ts
/**
 * Mirrors the JSON contract of POST /api/v1/query and GET /api/v1/databases
 * (see backend/internal/model/query.go).
 */

export interface QueryRequest {
  sql: string
  /** Scopes the statement (USE <db>) when set. */
  database?: string
  /** Requested row cap; the server clamps values above its maximum. */
  maxRows?: number
}

export interface QueryColumn {
  name: string
  /** Driver-reported database type, e.g. VARCHAR, BIGINT. May be empty. */
  type?: string
}

export interface QueryResult {
  columns: QueryColumn[]
  /** Cell values as strings, null for SQL NULL. */
  rows: (string | null)[][]
  rowCount: number
  /** True when the statement produced more rows than maxRows. */
  truncated: boolean
  /** The effective row cap applied to this execution. */
  maxRows: number
  /** Wall-clock execution time measured by the API server. */
  elapsedMs: number
  /** The cleaned SQL that was actually executed. */
  statement: string
  database?: string
}

export interface DatabasesResponse {
  databases: string[]
}

File: backend/internal/model/query.go
package model

// QueryRequest is the payload of POST /api/v1/query.
type QueryRequest struct {
	// SQL is a single statement; a trailing semicolon is tolerated.
	SQL string `json:"sql" binding:"required"`
	// Database scopes the statement (USE <db>) when set.
	Database string `json:"database"`
	// MaxRows caps the result set. Zero means the server default; values above
	// the server maximum are clamped, never honored.
	MaxRows int `json:"maxRows"`
}

// QueryColumn describes one result column.
type QueryColumn struct {
	Name string `json:"name"`
	// Type is the driver-reported database type name, e.g. VARCHAR, BIGINT.
	Type string `json:"type,omitempty"`
}

// QueryResult is the payload of a successful POST /api/v1/query.
type QueryResult struct {
	Columns []QueryColumn `json:"columns"`
	// Rows hold cell values as strings, with nil for SQL NULL. Everything is
	// read as text — the worksheet renders, it does not compute.
	Rows     [][]any `json:"rows"`
	RowCount int     `json:"rowCount"`
	// Truncated is true when the statement produced more rows than MaxRows.
	Truncated bool `json:"truncated"`
	// MaxRows echoes the effective row cap applied to this execution.
	MaxRows int `json:"maxRows"`
	// ElapsedMs is wall-clock execution time measured by the API server,
	// including result streaming — an operator-facing approximation, not the
	// engine's internal profile.
	ElapsedMs int64 `json:"elapsedMs"`
	// Statement is the cleaned SQL that was actually executed.
	Statement string `json:"statement"`
	Database  string `json:"database,omitempty"`
}
